### ISSUE TITLE
redirection in test generates SC1072/SC1073/SC1009

### ISSUE THREAD

**Issue by aldennisa15** [koalaman/shellcheck#1538]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC1072
- My shellcheck version (`shellcheck --version` or "online"): shellcheck.net
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

if ! [ "$1" -eq "$1" 2> /dev/null ]
then
  echo "ERROR: $1 is not a number!"
fi

```

#### Here's what shellcheck currently says:

```
Line 3:
if ! [ "$1" -eq "$1" 2> /dev/null ]
^-- SC1009: The mentioned syntax error was in this if expression.
     ^-- SC1073: Couldn't parse this test expression. Fix to allow more checks.
                     ^-- SC1072: Expected test to end here (don't wrap commands in []/[[]]). Fix any mentioned problems and try again.
````

#### Here's what I wanted or expected to see:

_Nothing_, it's valid?

(It's a test that $1 is an integer. Without the redirection to null, when executed, bash outputs: "./x.sh: line 3: [: a: integer expression expected" - that needs to be suppressed in favour of the echo error message instead,)

**Comment by aldennisa15** [comment]
I've since found that putting the redirection after the closing square bracket works fine. 

Maybe shellcheck was right all along... :-/

**Comment by mhoes** [comment]
If you want to test if a string is a number, you could also write it as the code below, which doesn't get you a syntax error if the tested string is not an integer :

```
#!/bin/bash

case $1 in
(*[!0-9]*|'')
        echo not a number
        ;;
*)
        echo a number
        ;;
esac
```

**Comment by aldennisa15** [comment]
Thanks. Indeed, there's probably dozens of ways to test for an integer, but that's a side issue. The point is, it wasn't clear if shellcheck was correctly objecting to the error redirection?

btw, that case test doesn't work for negative integers - in my experience, it's often safer to let established software do the tricky stuff, rather than trying to "reinvent the wheel" ;-)

**Comment by mhoes** [comment]
Point taken. 

As far as I can see, shellcheck was right to flag the code:

In the statement, the redirection is not part of the test (so not within [...]), but you want to redirect the output (stderr) of the test to /dev/null instead.

**Comment by koalaman** [comment]
You're right that this was a false positive. Your code was fine.

This is a conscious design decision though.  It helps catch errors like `[ 2>1 ]`, which is also a perfectly valid POSIX example of redirecting stderr inside `[ .. ]`, but no one would guess that this is what's going on.

Putting any actual redirections after the `]` is hopefully a small price to pay for improved diagnostics, and it doesn't change the meaning of the redirection.
